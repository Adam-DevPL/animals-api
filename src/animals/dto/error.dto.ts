import { ApiProperty } from '@nestjs/swagger';

export class ErrorDto {
  @ApiProperty({
    description: 'status of the error response',
    type: 'Number',
    example: '500',
  })
  status: number;

  @ApiProperty({
    description: 'error message',
    type: 'String',
    example: 'Internal server error',
  })
  message: string;
}
