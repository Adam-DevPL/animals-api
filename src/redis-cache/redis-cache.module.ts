import { CacheModule, Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { redisStore } from 'cache-manager-redis-store';
import { RedisCacheService } from './redis-cache.service';

@Module({
  imports: [
    CacheModule.register({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        store: redisStore,
        host: configService.get('REDIS_HOST'),
        port: configService.get('REDIS_PORT'),
      }),
    }),
  ],
  providers: [RedisCacheService],
})
export class RedisCacheModule {}
