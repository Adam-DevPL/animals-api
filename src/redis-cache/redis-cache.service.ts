import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';

@Injectable()
export class RedisCacheService {
  constructor(@Inject(CACHE_MANAGER) private readonly cacheManager: Cache) {}

  async clearAllCache() {
    this.cacheManager.del('findAll');
  }

  async clearSingleCache(id: string) {
    this.cacheManager.del(`/animals/animal/${id}`);
  }
}
