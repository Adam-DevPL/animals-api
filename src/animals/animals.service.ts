import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { RedisCacheService } from 'src/redis-cache/redis-cache.service';
import { AnimalTypeParam } from 'src/validations/type.validator';
import {
  AnimalDto,
  UpdateAnimalDto,
  AnimalNameDto,
  AnimalDtoResponse,
} from './dto/animal.dto';
import { Animal, AnimalDocument } from './schemas/animal.schema';

@Injectable()
export class AnimalsService {
  constructor(
    @InjectModel(Animal.name) private animalModel: Model<AnimalDocument>,
    private readonly redisCacheManager: RedisCacheService,
  ) {}

  async findAll(): Promise<AnimalDtoResponse[]> {
    const animals: Animal[] = await this.animalModel.find();
    return AnimalDtoResponse.mapperArrayDto(animals);
  }

  async findOne(id: string): Promise<AnimalDtoResponse> {
    const animal: AnimalDocument = await this.animalModel.findById({
      _id: new Types.ObjectId(id),
    });

    if (!animal) {
      throw new NotFoundException('Animal was not found in database');
    }

    return AnimalDtoResponse.mapperDto(animal);
  }

  async update(
    id: string,
    animalData: UpdateAnimalDto,
  ): Promise<AnimalDtoResponse> {
    const animal: AnimalDocument = await this.animalModel.findByIdAndUpdate(
      { _id: { $eq: new Types.ObjectId(id) } },
      animalData,
      {
        new: true,
      },
    );

    if (!animal) {
      throw new NotFoundException('Animal with given id does not exist');
    }

    await this.redisCacheManager.clearAllCache();
    await this.redisCacheManager.clearSingleCache(animal._id.toString());

    return AnimalDtoResponse.mapperDto(animal);
  }

  async create(animalData: AnimalDto): Promise<AnimalDtoResponse> {
    const animal: AnimalDocument = await this.animalModel.findOne({
      animalName: { $eq: animalData.animalName },
      animalType: { $eq: animalData.type },
    });

    if (animal) {
      throw new BadRequestException('Animal already exist in database');
    }

    const newAnimal: AnimalDocument = await this.animalModel.create({
      createdAt: new Date(),
      ...animalData,
    });

    await this.redisCacheManager.clearAllCache();

    return AnimalDtoResponse.mapperDto(newAnimal);
  }

  async addAnimals(animalsList: AnimalDto[]): Promise<AnimalDtoResponse[]> {
    const allAnimals: AnimalDtoResponse[] = await this.findAll();

    const alreadyExistAnimals: AnimalDto[] = animalsList.filter((animal) =>
      allAnimals.some(
        ({ name, type }) => animal.animalName === name && animal.type === type,
      ),
    );

    if (alreadyExistAnimals.length !== 0) {
      throw new BadRequestException('Animals already exist in database');
    }

    await this.redisCacheManager.clearAllCache();

    const animals: Animal[] = await this.animalModel.insertMany(
      animalsList.map((animal) => ({ createdAt: new Date(), ...animal })),
    );

    return AnimalDtoResponse.mapperArrayDto(animals);
  }
  async addAnimalsWithOneType(
    animalsNames: AnimalNameDto[],
    typeAnimal: AnimalTypeParam,
  ): Promise<AnimalDtoResponse[]> {
    const allAnimals: AnimalDtoResponse[] = await this.findAll();
    const alreadyExistAnimals: AnimalNameDto[] = animalsNames.filter((animal) =>
      allAnimals.some(
        ({ name, type }) =>
          animal.animalName === name && typeAnimal.type === type,
      ),
    );

    if (alreadyExistAnimals.length !== 0) {
      throw new BadRequestException('Animals already exist in database');
    }

    await this.redisCacheManager.clearAllCache();

    const animals: Animal[] = await this.animalModel.insertMany(
      animalsNames.map((animal) => ({
        createdAt: new Date(),
        animalName: animal.animalName,
        type: typeAnimal.type,
      })),
    );

    return AnimalDtoResponse.mapperArrayDto(animals);
  }
}
