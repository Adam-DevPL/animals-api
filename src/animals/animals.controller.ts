import { Body, Controller, Get, Param, Post, Put } from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiInternalServerErrorResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiOperation,
  ApiParam,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { ParamsWithId } from 'src/validations/id.validator';
import { AnimalsService } from './animals.service';
import { AnimalDto } from './dto/animal.dto';

@ApiTags('animals')
@Controller('animals')
export class AnimalsController {
  constructor(private readonly animalsService: AnimalsService) {}

  @Get('/all')
  @ApiOperation({ summary: 'Get all animals from the api' })
  @ApiResponse({ status: 200, description: 'Get all data' })
  @ApiResponse({ status: 500, description: 'Internal server error' })
  async getAllAnimals() {
    return this.animalsService.findAll();
  }

  @Get('animal/:id')
  @ApiParam({
    name: 'id',
    description: 'Gets the Animal id',
  })
  @ApiOperation({ summary: 'Get an animal with a given id' })
  @ApiOkResponse({ description: 'The resources were returned successfully' })
  @ApiBadRequestResponse({ description: 'Incorrect id number' })
  @ApiNotFoundResponse({ description: 'Animal not found' })
  @ApiInternalServerErrorResponse({ description: 'Internal server error' })
  async getAnimal(@Param() { id }: ParamsWithId) {
    return this.animalsService.findOne(id);
  }

  // @Post('/add')
  // async createAnimal(@Body() animal: AnimalDto) {
  //   return this.animalsService.create(animal);
  // }

  // @Put('animal/:id')
  // async updateAnimal(@Param() { id }: ParamsWithId, @Body() animal: AnimalDto) {
  //   return this.animalsService.update(id, animal);
  // }
}
