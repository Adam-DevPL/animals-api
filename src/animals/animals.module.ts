import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { RedisCacheModule } from 'src/redis-cache/redis-cache.module';
import { AnimalsController } from './animals.controller';
import { AnimalsService } from './animals.service';
import { Animal, AnimalSchema } from './schemas/animal.schema';

@Module({
  imports: [
    MongooseModule.forFeature([{ name: Animal.name, schema: AnimalSchema }]),
    RedisCacheModule,
  ],
  controllers: [AnimalsController],
  providers: [AnimalsService],
})
export class AnimalsModule {}
