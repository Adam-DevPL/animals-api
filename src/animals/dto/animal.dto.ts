import { Optional } from '@nestjs/common';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  IsString,
  IsNotEmpty,
  IsEnum,
  IsOptional,
  IsArray,
  ValidateNested,
} from 'class-validator';
import { AnimalType } from 'src/types/animals.type';

export class AnimalDto {
  @ApiProperty({
    description: 'Animal name, can not be empty',
    type: 'string',
    example: 'Cat',
  })
  @IsString()
  @IsNotEmpty()
  animalName: string;

  @ApiProperty({
    description: `One of Animal types - see Enum`,
    enum: AnimalType,
    type: 'AnimalType',
    example: AnimalType.MAMMALS,
  })
  @IsNotEmpty()
  @IsEnum(AnimalType)
  type: AnimalType;

  @ApiPropertyOptional({
    description: 'Optional description',
    type: 'string',
    example: 'test description',
  })
  @IsOptional()
  @IsString()
  description?: string;
}

export class UpdateAnimalDto {
  @ApiPropertyOptional({
    description: 'Animal name',
    type: 'string',
    example: 'Cat',
  })
  @IsOptional()
  @IsString()
  animalName?: string;

  @ApiPropertyOptional({
    description: `One of Animal types - see Enum`,
    enum: AnimalType,
    type: 'AnimalType',
    example: AnimalType.MAMMALS,
  })
  @Optional()
  @IsEnum(AnimalType)
  type?: AnimalType;

  @ApiPropertyOptional({
    description: 'Optional description',
    type: 'string',
    example: 'test description',
  })
  @IsOptional()
  @IsString()
  description?: string;
}

export class AnimalDtoArray {
  @ApiProperty({ description: 'Array with animals', type: [AnimalDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => AnimalDto)
  animals: AnimalDto[];
}
