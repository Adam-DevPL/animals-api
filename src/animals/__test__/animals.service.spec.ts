import {
  BadRequestException,
  CACHE_MANAGER,
  InternalServerErrorException,
  NotFoundException,
} from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';
import { Error, Model, Types } from 'mongoose';
import { RedisCacheService } from 'src/redis-cache/redis-cache.service';
import { AnimalType } from 'src/types/animals.type';
import { AnimalTypeParam } from 'src/validations/type.validator';
import { AnimalsService } from '../animals.service';
import { AnimalDto, AnimalNameDto, UpdateAnimalDto } from '../dto/animal.dto';
import { Animal, AnimalDocument } from '../schemas/animal.schema';

const animalStub: Animal = {
  _id: new Types.ObjectId('507f1f77bcf86cd799439011'),
  animalName: 'Human',
  type: AnimalType.MAMMALS,
  createdAt: new Date('2023-03-19T18:27:12.933Z'),
  description: 'test description',
};

const animalUpdateDto: UpdateAnimalDto = {
  animalName: 'Cat',
  type: AnimalType.BIRDS,
  description: 'description after update',
};

const animalDto: AnimalDto = {
  animalName: 'Dog',
  type: AnimalType.MAMMALS,
  description: 'new dog created',
};

const animalsList: AnimalDto[] = [
  {
    animalName: 'Elephant',
    type: AnimalType.MAMMALS,
  },
  {
    animalName: 'Lion',
    type: AnimalType.MAMMALS,
  },
];

const animalsNamesList: AnimalNameDto[] = [
  {
    animalName: 'Cat',
  },
  {
    animalName: 'Dog',
  },
];

describe('AnimalsService', () => {
  let service: AnimalsService;
  let model: Model<AnimalDocument>;
  let redisCache: RedisCacheService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        RedisCacheService,
        AnimalsService,
        {
          provide: CACHE_MANAGER,
          useValue: {
            del: () => 'any value',
          },
        },
        {
          provide: getModelToken(Animal.name),
          useValue: {
            find: jest.fn().mockResolvedValue([animalStub]),
            findOne: jest.fn().mockResolvedValue(null),
            findById: jest
              .fn()
              .mockImplementation((id: string) =>
                Promise.resolve({ _id: id, ...animalStub }),
              ),
            findByIdAndUpdate: jest
              .fn()
              .mockImplementation((id: string, animalData: UpdateAnimalDto) =>
                Promise.resolve({
                  _id: id,
                  createdAt: new Date('2023-03-19T18:27:12.933Z'),
                  ...animalData,
                }),
              ),
            create: jest.fn().mockImplementation((animalData: AnimalDto) =>
              Promise.resolve({
                _id: '507f1f77bcf86cd799439011',
                ...animalData,
              }),
            ),
            insertMany: jest.fn().mockImplementation((animals: AnimalDto[]) =>
              Promise.resolve([
                {
                  _id: 'id1',
                  ...animals[0],
                },
                {
                  _id: 'id2',
                  ...animals[1],
                },
              ]),
            ),
          },
        },
      ],
    }).compile();

    service = module.get<AnimalsService>(AnimalsService);
    model = module.get<Model<AnimalDocument>>(getModelToken(Animal.name));
    redisCache = module.get<RedisCacheService>(RedisCacheService);
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('findAll', () => {
    it('should find all animals', async () => {
      //when
      const result = await service.findAll();
      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw Internal Exception Error when something goes wrong with mongodb connection', async () => {
      //when
      jest.spyOn(model, 'find').mockImplementation(() => {
        throw new Error('mongodb down');
      });

      //then
      await expect(async () => {
        await service.findAll();
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });

  describe('findOne', () => {
    it('should return one animal', async () => {
      //given
      const id = '507f1f77bcf86cd799439011';

      //when
      const result = await service.findOne(id);

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error NotFoundException --> Not Found, when animal does not exist', async () => {
      //when
      jest.spyOn(model, 'findById').mockImplementation(() => {
        throw new NotFoundException();
      });

      //then
      await expect(async () => {
        await service.findOne('507f1f77bcf86cd799439099');
      }).rejects.toThrowError(NotFoundException);
    });

    it('should throw error InternalServerErrorException when mongodb fails', async () => {
      //when
      jest.spyOn(model, 'findById').mockImplementation(() => {
        throw new Error('mongo fails');
      });

      //then
      await expect(async () => {
        await service.findOne('507f1f77bcf86cd799439099');
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });

  describe('update', () => {
    it('should return updated animal', async () => {
      //given
      const id = '507f1f77bcf86cd799439011';

      //when
      const result = await service.update(id, animalUpdateDto);

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error NotFoundException --> Not Found, when animal does not exist', async () => {
      //when
      jest.spyOn(model, 'findByIdAndUpdate').mockImplementation(() => {
        throw new NotFoundException();
      });

      //then
      await expect(async () => {
        await service.update('507f1f77bcf86cd799439099', animalUpdateDto);
      }).rejects.toThrowError(NotFoundException);
    });

    it('should throw error InternalServerErrorException when mongodb fails', async () => {
      //when
      jest.spyOn(model, 'findByIdAndUpdate').mockImplementation(() => {
        throw new Error('mongo fails');
      });

      //then
      await expect(async () => {
        await service.update('507f1f77bcf86cd799439099', animalUpdateDto);
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });

  describe('create', () => {
    it('should return newly created animal', async () => {
      //when
      const result = await service.create(animalDto);

      //then
      expect({
        ...result,
        createdAt: new Date('2022-02-02'),
      }).toMatchSnapshot();
    });

    it('should throw error BadRequestException --> Animal already exist in database', async () => {
      //when
      jest.spyOn(model, 'findOne').mockResolvedValue({ _id: 'id' });

      //then
      await expect(async () => {
        await service.create(animalDto);
      }).rejects.toThrowError(BadRequestException);
    });

    it('should throw error InternalServerErrorException when mongodb fails', async () => {
      //when
      jest.spyOn(model, 'findOne').mockImplementation(() => {
        throw new Error('mongo fails');
      });

      //then
      await expect(async () => {
        await service.create(animalDto);
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });

  describe('addAnimals', () => {
    it('should return newly created animals', async () => {
      //when
      const result = (await await service.addAnimals(animalsList)).map(
        (res) => ({
          createdAt: (res.createdAt = '2022-02-02'),
          ...res,
        }),
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error BadRequestException --> Animal from the list already exist in db', async () => {
      //when
      jest.spyOn(service, 'findAll').mockImplementation(() =>
        Promise.resolve([
          {
            id: 'id1',
            name: 'Elephant',
            type: AnimalType.MAMMALS,
            createdAt: '2023-03-19T18:27:12.933Z',
            description: 'test description',
          },
          {
            id: 'id2',
            name: 'Lion',
            type: AnimalType.MAMMALS,
            createdAt: '2023-03-19T18:27:12.933Z',
            description: 'test description',
          },
        ]),
      );

      //then
      await expect(async () => {
        await service.addAnimals(animalsList);
      }).rejects.toThrowError(BadRequestException);
    });

    it('should throw error InternalServerErrorException when mongodb fails', async () => {
      //when
      jest.spyOn(model, 'insertMany').mockImplementation(() => {
        throw new Error('mongo fails');
      });

      //then
      expect(async () => {
        await service.addAnimals(animalsList);
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });

  describe('addAnimalsWithOneType', () => {
    it('should return newly created animals', async () => {
      //given
      const type: AnimalTypeParam = { type: AnimalType.MAMMALS };

      //when
      const result = (
        await service.addAnimalsWithOneType(animalsNamesList, type)
      ).map((res) => ({
        createdAt: (res.createdAt = '2022-02-02'),
        ...res,
      }));

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error BadRequestException --> Animal from the list already exist in db', async () => {
      //given
      const type: AnimalTypeParam = { type: AnimalType.MAMMALS };

      //when
      jest.spyOn(service, 'findAll').mockImplementation(() =>
        Promise.resolve([
          {
            id: 'id1',
            name: 'Cat',
            type: AnimalType.MAMMALS,
            createdAt: '2023-03-19T18:27:12.933Z',
            description: 'test description',
          },
          {
            id: 'id2',
            name: 'Dog',
            type: AnimalType.MAMMALS,
            createdAt: '2023-03-19T18:27:12.933Z',
            description: 'test description',
          },
        ]),
      );

      //then
      await expect(async () => {
        await service.addAnimalsWithOneType(animalsNamesList, type);
      }).rejects.toThrowError(BadRequestException);
    });

    it('should throw error InternalServerErrorException when mongodb fails', async () => {
      //given
      const type: AnimalTypeParam = { type: AnimalType.MAMMALS };

      //when
      jest.spyOn(model, 'insertMany').mockImplementation(() => {
        throw new Error('mongo fails');
      });

      //then
      await expect(async () => {
        await service.addAnimalsWithOneType(animalsNamesList, type);
      }).rejects.toThrowError(InternalServerErrorException);
    });
  });
});
