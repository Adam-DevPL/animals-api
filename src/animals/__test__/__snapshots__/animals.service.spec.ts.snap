// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalsService findAll should find all animals 1`] = `
[
  {
    "animalName": "Human",
    "createdAt": 2023-03-19T18:27:12.933Z,
    "description": "test description",
    "type": "MAMMALS",
  },
]
`;

exports[`AnimalsService findAll should throw Internal Exception Error when something goes wrong with mongodb connection 1`] = `[InternalServerErrorException: Internal Server Error]`;

exports[`AnimalsService findOne should return one animal 1`] = `
{
  "animalName": "Human",
  "createdAt": 2023-03-19T18:27:12.933Z,
  "description": "test description",
  "type": "MAMMALS",
}
`;

exports[`AnimalsService findOne should throw error NotFoundException --> Not Found, when animal does not exist 1`] = `[NotFoundException: Not Found]`;
